File: gateway/emitter.py
from __future__ import annotations

from shared.agent_io import EmitRequest
from shared.db.client import load_agent_session, load_card_context
from shared.logging import logger

from .channel_registry import ChannelRegistry
from .models import OutboundRequest


class GatewayEmitter:
    def __init__(self, *, registry: ChannelRegistry) -> None:
        self._registry = registry

    async def emit(self, emit: EmitRequest) -> None:
        session = await load_agent_session(emit.session_id)
        if session is None:
            raise RuntimeError(f"agent session not found: {emit.session_id}")
        emit_kind = str(emit.emit_kind or "").strip()
        card_id = str(emit.card_id or "").strip()
        card_ctx = await load_card_context(card_id) if card_id else None
        source = dict(getattr(session, "source", {}) or {})
        platform = str(card_ctx.platform or "").strip() if card_ctx is not None else str(source.get("platform") or "").strip()
        adapter = self._registry.get(platform)
        logger.info(
            "[GatewayEmitter] dispatch emit: session_id=%s kind=%s platform=%s card_id=%s content_len=%d files=%d",
            emit.session_id,
            emit_kind,
            platform,
            card_id,
            len(str(emit.content or "")),
            len(list(emit.files or [])),
        )
        if emit_kind == "progress":
            logger.info("[GatewayEmitter] ignore progress emit: session_id=%s emit_id=%s", emit.session_id, emit.emit_id)
            return
        if emit_kind == "stream":
            await self._send_stream_message(adapter, emit=emit, platform=platform, card_id=card_id)
            return
        if emit_kind not in {"tool", "final"}:
            raise RuntimeError(f"unsupported emit_kind: {emit_kind}")

        if emit_kind == "tool":
            await self._send_files(adapter, emit=emit, platform=platform, card_id=card_id)
            if emit.content:
                await self._send_message(adapter, emit=emit, platform=platform, card_id=card_id, content=emit.content)
            return

        if emit.content:
            await self._send_message(adapter, emit=emit, platform=platform, card_id=card_id, content=emit.content)
        await self._send_files(adapter, emit=emit, platform=platform, card_id=card_id)

    async def emit_stream(
        self,
        *,
        session_id: str,
        card_id: str,
        stream_type: str,
        state: str,
        seq: int,
        content: str,
        emit_id: str = "",
    ) -> None:
        await self.emit(
            EmitRequest(
                session_id=str(session_id or "").strip(),
                card_id=str(card_id or "").strip(),
                content=str(content or "").strip(),
                emit_kind="stream",
                emit_id=str(emit_id or "").strip(),
                stream_type=str(stream_type or "").strip(),
                state=str(state or "").strip(),
                seq=int(seq or 0),
            )
        )

    async def emit_final(
        self,
        *,
        session_id: str,
        card_id: str = "",
        content: str,
        files: list[str] | None = None,
        emit_id: str = "",
    ) -> None:
        await self.emit(
            EmitRequest(
                session_id=str(session_id or "").strip(),
                card_id=str(card_id or "").strip(),
                content=str(content or "").strip(),
                files=list(files or []),
                emit_kind="final",
                emit_id=str(emit_id or "").strip(),
            )
        )

    async def _send_stream_message(
        self,
        adapter,
        *,
        emit: EmitRequest,
        platform: str,
        card_id: str,
    ) -> None:
        safe_content = str(emit.content or "").strip()
        if not safe_content:
            return
        logger.info(
            "[GatewayEmitter] stream_message: session_id=%s card_id=%s stream_type=%s state=%s seq=%d content_len=%d",
            emit.session_id,
            card_id,
            str(emit.stream_type or "").strip(),
            str(emit.state or "").strip(),
            int(emit.seq or 0),
            len(safe_content),
        )
        request = OutboundRequest(
            action="stream_message",
            platform=platform,
            payload={
                "stream_type": str(emit.stream_type or "").strip(),
                "state": str(emit.state or "").strip(),
                "seq": int(emit.seq or 0),
                "content": safe_content,
            },
            session_id=emit.session_id,
            card_id=card_id,
            event_id=emit.emit_id,
        )
        await adapter.handle_outbound(request)

    async def _send_message(
        self,
        adapter,
        *,
        emit: EmitRequest,
        platform: str,
        card_id: str,
        content: str,
    ) -> None:
        safe_content = str(content or "").strip()
        if not safe_content:
            return
        logger.info("[GatewayEmitter] send_message: session_id=%s card_id=%s content_len=%d", emit.session_id, card_id, len(safe_content))
        request = OutboundRequest(
            action="send_message",
            platform=platform,
            payload={"markdown": safe_content},
            session_id=emit.session_id,
            card_id=card_id,
            event_id=emit.emit_id,
        )
        await adapter.handle_outbound(request)

    async def _send_files(
        self,
        adapter,
        *,
        emit: EmitRequest,
        platform: str,
        card_id: str,
    ) -> None:
        for path in list(emit.files or []):
            file_path = str(path or "").strip()
            if not file_path:
                continue
            logger.info("[GatewayEmitter] send_file: session_id=%s card_id=%s path=%s", emit.session_id, card_id, file_path)
            request = OutboundRequest(
                action="send_file",
                platform=platform,
                payload={"path": file_path},
                session_id=emit.session_id,
                card_id=card_id,
                event_id=emit.emit_id,
            )
            await adapter.handle_outbound(request)


__all__ = ["GatewayEmitter"]

File: apps/gateway/src/emitter.ts
import { randomUUID } from "node:crypto";
import {
  validateEmitRequest,
  type EmitRequest,
  type JsonObject,
} from "@lxe/protocol";
import type { ChannelRegistry } from "./channel";
import type { OutboundRequest, ResponseRouteRecord } from "./models";

interface EmitterRoutePort {
  getSession(sessionId: string): Promise<{ session_id: string; source: JsonObject } | undefined>;
  getResponseRoute(responseRouteId: string): Promise<ResponseRouteRecord | undefined>;
}

export interface GatewayEmitterOptions {
  registry: ChannelRegistry;
  routes: EmitterRoutePort;
  uuid?: () => string;
}

const safeText = (value: unknown): string => String(value ?? "").trim();

export class GatewayEmitter {
  private readonly uuid: () => string;

  constructor(private readonly options: GatewayEmitterOptions) {
    this.uuid = options.uuid ?? (() => randomUUID().replaceAll("-", ""));
  }

  async emit(emit: EmitRequest): Promise<void> {
    if (!validateEmitRequest(emit)) {
      throw new Error(`invalid runtime.emit payload: ${validateEmitRequest.errors?.[0]?.message ?? "invalid request"}`);
    }
    const context = await this.resolve(emit.session_id, emit.response_route_id);
    if (!context) throw new Error("response context unexpectedly unavailable");
    const adapter = this.options.registry.get(context.platform);
    const kind = safeText(emit.emit_kind);
    if (kind === "progress") return;
    if (kind === "stream") {
      const content = safeText(emit.content);
      const thinking = safeText(emit.thinking);
      if (!content && !thinking && emit.redacted_thinking_count <= 0 && !emit.tool_pending && emit.tool_steps.length === 0) return;
      await adapter.handleOutbound(this.request(emit, context.platform, "stream_message", {
        stream_type: safeText(emit.stream_type),
        state: safeText(emit.state),
        seq: emit.seq,
        content,
        thinking,
        redacted_thinking_count: emit.redacted_thinking_count,
        thinking_elapsed_ms: emit.thinking_elapsed_ms,
        tool_pending: emit.tool_pending,
        tool_elapsed_ms: emit.tool_elapsed_ms,
        tool_steps: emit.tool_steps.map((item) => ({ ...item })),
        display_metrics: { ...emit.display_metrics },
      }));
      return;
    }
    if (!new Set(["tool", "final"]).has(kind)) throw new Error(`unsupported emit_kind: ${kind}`);
    if (kind === "tool") {
      await this.sendFiles(adapter, emit, context.platform);
      await this.sendMessage(adapter, emit, context.platform);
      return;
    }
    await this.sendMessage(adapter, emit, context.platform);
    await this.sendFiles(adapter, emit, context.platform);
  }

  async typing(payload: {
    session_id: string;
    response_route_id: string;
    operation: string;
    emit_id?: string;
  }): Promise<void> {
    const sessionId = safeText(payload.session_id);
    const responseRouteId = safeText(payload.response_route_id);
    const operation = safeText(payload.operation);
    const emitId = safeText(payload.emit_id);
    if (!sessionId || !responseRouteId) throw new Error("invalid runtime.typing payload");
    if (!new Set(["start", "stop"]).has(operation)) throw new Error(`unsupported typing operation: ${operation}`);
    const context = await this.resolve(sessionId, responseRouteId, true);
    if (!context || context.platform !== "feishu") return;
    await this.options.registry.get(context.platform).handleOutbound({
      action: "typing_indicator",
      platform: context.platform,
      payload: { operation },
      session_id: sessionId,
      response_route_id: responseRouteId,
      event_id: emitId || this.uuid(),
    });
  }

  private async resolve(
    sessionId: string,
    responseRouteId: string,
    allowMissingRoute = false,
  ): Promise<{ platform: string; route?: ResponseRouteRecord } | undefined> {
    const session = await this.options.routes.getSession(safeText(sessionId));
    if (!session) throw new Error(`agent session not found: ${sessionId}`);
    const safeRouteId = safeText(responseRouteId);
    const route = safeRouteId ? await this.options.routes.getResponseRoute(safeRouteId) : undefined;
    if (safeRouteId && !route) {
      if (allowMissingRoute) return undefined;
      throw new Error(`response route not found: ${safeRouteId}`);
    }
    const platform = safeText(route?.platform) || safeText(session.source.platform);
    if (!platform) throw new Error(`platform not found for session: ${sessionId}`);
    return { platform, ...(route ? { route } : {}) };
  }

  private request(emit: EmitRequest, platform: string, action: string, payload: JsonObject): OutboundRequest {
    return {
      action,
      platform,
      payload,
      session_id: emit.session_id,
      response_route_id: emit.response_route_id,
      event_id: emit.emit_id,
    };
  }

  private async sendMessage(
    adapter: { handleOutbound(request: OutboundRequest): Promise<void> },
    emit: EmitRequest,
    platform: string,
  ): Promise<void> {
    const content = safeText(emit.content);
    if (!content) return;
    await adapter.handleOutbound(this.request(emit, platform, "send_message", { markdown: content }));
  }

  private async sendFiles(
    adapter: { handleOutbound(request: OutboundRequest): Promise<void> },
    emit: EmitRequest,
    platform: string,
  ): Promise<void> {
    for (const value of emit.files) {
      const path = safeText(value);
      if (path) await adapter.handleOutbound(this.request(emit, platform, "send_file", { path }));
    }
  }
}
